client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -120,10 +125,12 @@ public synchronized void setHosts(HttpHost... hosts) {
         if (hosts == null || hosts.length == 0) {
             throw new IllegalArgumentException("hosts must not be null nor empty");
         }
+        authCache.clear();
         Set<HttpHost> httpHosts = new HashSet<>();
         for (HttpHost host : hosts) {
             Objects.requireNonNull(host, "host cannot be null");
             httpHosts.add(host);
+            authCache.put(host, new BasicScheme());
```

Comment:
> but there is still the issue that there will be a small amount of time when hosts and the auth cache are not in sync (between setting both).

I'll make them a `volatile Tuple` that get updated together. That way there's no opportunity for a dirty read.